Spring Security 적용

BEOMKING opened the pull request:
## Summary
`Spring Security`를 이용해서 보안을 적용합니다.

## Detail
- 유저에 일반 유저와 관리자를 구분하는 `Role`을 적용합니다.
- 상품 정보 변경은 상품 주인과 관리자만 가능합니다.

BEOMKING committed:
인터셉터 방식 ->  Spring Security JWT 필터로 변경

BEOMKING committed:
Filter 적용

Commit:
불필요한 코드 제거

Hello 컨트롤러 제거
User 컨트롤러 테스트 재구현을 위해 제거

Commit:
유저 재구현

Commit:
유저 컨트롤러 생성 테스트 작성

Commit:
유저 서비스 등록 테스트 작성

hannut91 commented on app/src/main/java/com/codesoom/assignment/interceptors/JwtAuthenticationFilter.java:
```diff
@@ -0,0 +1,68 @@
+        chain.doFilter(request, response);
+    }
+
+    private boolean filterWithPathAndMethod(HttpServletRequest request) {
+        String path = request.getRequestURI();
+        if (!path.startsWith("/products")) {
+            return true;
+        }
+
+        String method = request.getMethod();
+        if (method.equals("GET")) {
+            return true;
+        }
+
+        if (method.equals("OPTIONS")) {
+            return true;
+        }
+
+        return false;
+    }
```
여기서 true를 반환할 때만 인증을 확인하고 있는 것 같네요. 나중에 애플리케이션이 점점 거대해지면 여기에 계속 추가해야 되겠네요

hannut91 reviewed (commented):
마지막까지 화이팅입니다!

Commit:
유저 생성 구현

BEOMKING committed:
중복된 이메일이 주어진 경우 예외 처리 구현

BEOMKING committed:
상품 로직 제거

BEOMKING committed:
프로젝트 생성 인수 테스트 작성

BEOMKING committed:
상품 생성 구현

유저의 id를 입력에 받아 상품을 생성합니다.

Commit:
토큰 암호화 클레임 정보 추가

Commit:
유저 유효한 로그인 구현

Commit:
테스트 중 생성된 유저에 다른 테스트가 영향을 받는 문제 제

hannut91 commented on app/src/main/java/com/codesoom/assignment/application/NormalUserService.java:
```diff
@@ -0,0 +1,27 @@
+
+    @Override
+    public UserInquiryInfo register(UserRegisterData registerData) {
+        if (userRepository.existsByEmail(registerData.getEmail())) {
+            throw new DuplicatedEmailException(registerData.getEmail());
+        }
```
이메일이 중복되면 안되나 보네요. 데이터 베이스에 email을 중복으로 저장할 수 없도록 할 수도 있겠어요

BEOMKING replied:
질문이 있습니다!

1. JPA 를 사용할 때는 `Entity`에 `Unique` 를 지정해주어 제약 조건을 쉽게 알 수 있습니다. 
그런데 `Unique` 제약 조건을 걸었어도 위 코드처럼 코드 레벨의 검증을 하시는 편이신가요? DB의 검증과 코드의 검증을 따로따로 하는 것이 선택의 영역인지 궁금하네요.

2. 만약 `MyBatis` 처럼 객체에 제약 조건을 명시할 수 없는 상황이라면 개발자가 DB의 제약 조건을 알기가 힘들다고 생각합니다. 이런 경우는 DB에 제약 조건을 설정해놓았더라도 명시적으로 코드 레벨의 검증을 해주는 것이 좋을 것 같은데 어떻게 생각하시나요?

hannut91 replied:
1. `Unique`제약 조건을 걸었다면 저는 검증하지 않을 것 같아요. DB가 잘하는 일은 DB가 하게 할 것 같습니다.
2. 개발자가 DB의 제약 조건을 알기 어려울 수 있겠네요. 그런 경우에는 명시적으로 코드 레벨에서 검증해줄 수 있는 방법이 좋겠네요.

지금은 단순하게 유일한 값이 들어가도록 하는 제약 조건에 대해서만 다루고 있다보니 데이터베이스가 처리해주는게 맞는 것 같은데, 데이터의 일관성은 검증은 결국 DB가 아니라 애플리케이션에서 할 수 밖에 없습니다.

hannut91 commented on app/src/test/java/com/codesoom/assignment/controllers/UserControllerTest.java:
```diff
@@ -1,157 +1,97 @@
+        @Nested
+        @DisplayName("중복되는 이메일이 주어지면")
+        class Context_with_email {
+            void prepare() throws Exception {
+                createUser(USER_DATA);
+            }
```
테스트를 위해서 준비하는 코드인 것 같네요. `@BeforeEach`를 사용하는 것은 어떨까요?

```java
@BeforeEach
void prepare() {
    createUser(USER_DATA);
}
```

BEOMKING replied:
준비 코드가 실행 공간에 들어가지 않아서 더 좋겠네요!

hannut91 commented on app/src/test/java/com/codesoom/assignment/controllers/ProductControllerTest.java:
```diff
@@ -1,268 +1,83 @@
+    public static final String PRICE = "10000";
+    public static final String USER_PATH = "/users";
+    public static final String PRODUCT_PATH = "/products";
+    public static final String EMAIL = "qjawlsqjacks@naver.com";
+    public static final String PASSWORD = "1234";
+    public static final String USER_NAME = "박범진";
```
테스트를 위한 사용자 데이터를 만드는 것이 여러 테스트에서 보이네요. 믿을 수 있는 데이터를 미리 만들어놓고 가져와서 사용하는 것은 어떨까요?

BEOMKING replied:
`Fixture`를 활용해보겠습니다.

Commit:
제품 생성할 때 토큰이 필요하도록 테스트 변경

BEOMKING committed:
로그인 테스트 실패 수정

준비 컨텍스트에 회원 가입 추가

BEOMKING committed:
테스트 코드 리팩터링

세션 테스트 테스트 고립성(테스트에 사용한 유저 제거)
자주 사용하는 필드 Fixture 재사용

BEOMKING committed:
테스트 코드 리팩터링(Fixture 사용)

hannut91 commented on app/src/test/java/com/codesoom/assignment/controllers/UserControllerTest.java:
```diff
@@ -1,157 +1,95 @@
-                .andExpect(status().isNoContent());
-
-        verify(userService).deleteUser(1L);
+    @AfterEach
+    void tearDown() {
+        userRepository.deleteAll();
     }
```
각 테스트가 끝나고 데이터 삭제해주고 있는 것 같아요. 만약 어떤 테스트가 도중에 실패해서 이 로직이 실행되지 않을 경우에 다른 테스트에 영향이 갈 수 있을 것 같아요.

테스트가 시작하기 전에 초기화를 하고 시작하는 것은 어떨까요?

BEOMKING replied:
도중에 실패할 수 있는 경우가 있었군요!

BEOMKING replied:
그런데 이런 경우가 있었습니다.

상품 서비스가 먼저 테스트가 되었을 경우, 유저와 상품이 하나씩 생성되어 있던 채로 서비스 테스트가 종료됩니다.
이후 유저 서비스가 테스트가 시작될 때, `BeforeEach` 를 사용해 유저를 지우는 로직에서 외래키와 관련된 에러가 발생하였습니다.

이유는 유저와 연관된 상품이 외래키로 연결되어 있어 유저를 삭제할 수 없는 경우였습니다.

제가 생각한 경우는
- 상품 테스트처럼 유저 테스트에도 상품 DB를 지우고 유저를 지운다.
- Entity에 Cascade 설정을 한다.
두 가지가 있습니다.

첫 번째 경우는 순수한 유저 테스트 로직에 상품 DB가 들어온다는 단점이 있고
두 번째 경우는 제약 조건 요구사항이 Cascade delete 설정이 아닐 때는 사용할 수 없을 것 같습니다.

이런 상황을 겪으신 경험이 있으신가요?

hannut91 replied:
데이터를 초기화하는 부분에서 어려움을 겪으셨네요. 생각해보니 저도 예전에 그런 경험이 있었네요. 그래서 저는 유저 테스트 로직에도 다른 테이블을 초기화하고 나서 초기호를 했던 경험이 있네요.

테스트가 시작되기 전에 테스트용 데이터베이스는 완전히 초기 상태에서 시작되면 좋겠네요. 이런 방법이 있을까요

BEOMKING replied:
`@Sql` DB를 초기화하거나 `@Transactional`으로 매 테스트에서 생긴 데이터를 롤백하면 좋을 것 같아요!

BEOMKING committed:
인증 토큰에 권한 추가

BEOMKING committed:
상품 등록 유저를 조회하는 로직 변경

상품 등록시 유저를 조회하는 로직을 DTO 에서 하는 로직 ->
컨텍스트에 있는 토큰에서 유저 정보를 가져와서 처리하는 로직으로 변경했습니다.

이 과정에서 Product에 userId 를 저장하는 구조 -> User 를 저장하도록 변경했습니다.

hannut91 commented on app/src/test/java/com/codesoom/assignment/controllers/ProductControllerTest.java:
```diff
@@ -1,268 +1,94 @@
+    @Nested
+    @DisplayName("register 메서드는")
+    class Describe_register {
+        @Nested
+        @DisplayName("상품 정보와 유효한 토큰이 주어지면")
+        class Context_userData {
```
Context는 주어진 상황이 달라지는 경우를 명시하는데, 유효한 토큰이 주어진 경우 명시한 이유는 아마도 토크인 주어지지 않은 경우를 염두해 두고 작성해주신 것 같아요.

토큰이 유효하지 않은 경우도 추가해보는 것은 어떨까요

hannut91 commented on app/src/test/java/com/codesoom/assignment/application/AuthenticationServiceTest.java:
```diff
@@ -1,87 +1,72 @@
+            @BeforeEach
+            void prepare() {
+                requestData = new SessionRequestData(Fixture.EMAIL, Fixture.PASSWORD);
+                given(userRepository.findByEmail(Fixture.EMAIL))
```
여기 위에 빈 줄 하나를 추가해서 로그인 요청을 위한 데이터를 생성하는 것과 모킹하는 것을 구분하면 좋겠어요

hannut91 commented on app/src/test/java/com/codesoom/assignment/application/AuthenticationServiceTest.java:
```diff
@@ -1,87 +1,72 @@
+            void It_returns_createdToken() {
+                SessionResponseData token = authenticationService.login(requestData);
+
+                Claims decode = jwtUtil.decode(token.getAccessToken());
+
+                assertThat(decode)
```
토큰을 decode한 결과라서 변수의 이름이 `decode`인 것 같아요. 결과 자체는 claim이니 변수 이름을 `claim`이라고 하는 것은 어떨까요

hannut91 commented on app/src/test/java/com/codesoom/assignment/Fixture.java:
```diff
@@ -0,0 +1,20 @@
+/**
+ * 고양이 장난감 가게 서비스 테스트에 사용되는 데이터 저장소입니다.
+ */
+public class Fixture {
```
자주 사용되는 문자열만 모여있는 것 같은데, 데이터 자체를 저장해놓는 것은 어떨까요? 사용자라던지 상품이라던지

Commit:
이메일 Unique 설정

Commit:
BeforeEach 적용, 무한 루프 해결

무한 루프 해결을 위해 ProductInquiryInfo라는 DTO를 생성해서 등록 메서드
시그니처를 변경했습니다.

Commit:
테스트 코드 가독성을 위해 줄 추가

hannut91 commented on app/src/main/java/com/codesoom/assignment/application/AuthenticationService.java:
```diff
@@ -12,25 +14,30 @@ public class AuthenticationService {
     }
 
-    public String login(String email, String password) {
+    /**
+     * 유효한 로그인이라면 토큰을 생성하고 리턴한다.
+     *
+     * @param data 로그인 정보
+     * @return 토큰 정보
+     */
```
`@param`에 변수명보다는 타입을 명시하는게 더 맞지 않나? 의문이 들어서, 가이드를 찾아 보니 변수명을 적고 설명에서 타입을 설명하라고 하네요.

> The @param tag is followed by the name (not data type) of the parameter, followed by a description of the parameter. By convention, the first noun in the description is the data type of the parameter. (Articles like "a", "an", and "the" can precede the noun.) An exception is made for the primitive int, where the data type is usually omitted. Additional spaces can be inserted between the name and description so that the descriptions line up in a block. Dashes or other punctuation should not be inserted before the description, as the Javadoc tool inserts one dash.
> 
> Parameter names are lowercase by convention. The data type starts with a lowercase letter to indicate an object rather than a class. The description begins with a lowercase letter if it is a phrase (contains no verb), or an uppercase letter if it is a sentence. End the phrase with a period only if another phrase or sentence follows it.

See also
  - https://www.oracle.com/technical-resources/articles/java/javadoc-tool.html

hannut91 commented on app/src/main/java/com/codesoom/assignment/application/NormalProductService.java:
```diff
@@ -0,0 +1,34 @@
+import org.springframework.stereotype.Service;
+
+@Service
+public class NormalProductService implements ProductService {
```
`Normal`이라는 것이 이해하기 어렵네요. 일반 상품이라는 것이 어떤 상품인가요?

BEOMKING replied:
인터페이스를 활용하고 싶어서 구현체를 만들었는데 마땅한 이름이 생각나지 않았고 Impl을 붙히기는 싫어서 만들어진 임시 이름이라고도 할 수 있습니다.

BEOMKING committed:
상품 생성시 유효하지 않은 토큰이 왔을 경우

BEOMKING committed:
유저 컨트롤러 삭제 인수 테스트 작성

BEOMKING committed:
유저 삭제 기능 구현

BEOMKING committed:
유저 관리자 권한이 아닐 때 삭제 금지 메서드 구현

hannut91 commented on app/src/test/java/com/codesoom/assignment/controllers/UserControllerTest.java:
```diff
@@ -1,157 +1,144 @@
+        }
+
+        @Nested
+        @DisplayName("유저 권한을 가진 토큰과 식별자가 주어지면")
```
유저 권한을 가지고 삭제를 요청을 할 경우 거부된다는 테스트를 하고 있는 것 같아요. `삭제 권한이 없는 사용자가 요청을 하면`이라고 하는 것도 나쁘지 않을 것 같네요

hannut91 commented on app/src/test/java/com/codesoom/assignment/controllers/UserControllerTest.java:
```diff
@@ -1,157 +1,144 @@
+            @Test
+            @DisplayName("유저와 201을 응답한다")
+            void It_returns_user() throws Exception {
+                mockMvc.perform(post(Fixture.USER_PATH)
```
요청 경로를 `Fixture`라고 볼 수 있을지 고민이 되네요. 제가 생각하기에는 `/users`라고 하는게 테스트를 읽는 입장에서는 더 이해가 쉬울 것 같은데, 나중에 경로가 변경될 경우 테스트에 영향을 가는 것을 고려하면 지금도 괜찮을 것 같아요

BEOMKING replied:
읽는 입장에서는 확실히 의문이 생길 수 있을 것 같아요. 장단점이 있군요!

hannut91 commented on app/src/test/java/com/codesoom/assignment/controllers/ProductControllerTest.java:
```diff
@@ -1,268 +1,120 @@
+            }
+
+            @Test
+            @DisplayName("상품을 생성하고 상품 조회 정보와 201을 응답한다")
```
401 Unauthorized를 응답해야 겠네요
